### ISSUE TITLE
Staging currently down: Memory quota exceeded

### ISSUE THREAD

**Issue by SimonLab** [club-soda/club-soda-guide#315]
While trying to run a seed file on staging the following error occurs:
![image](https://user-images.githubusercontent.com/6057298/50822402-94a5ee00-1329-11e9-8478-ef0610d429d7.png)

I think it is link to the ETS/geolcation feature but needs to look in a bit more detail to confirm the origin of this error.

It migth also just be the case that the Heroku dyno will need to be upgraded.
see https://devcenter.heroku.com/articles/ruby-memory-use

**Comment by RobStallion** [comment]
@SimonLab Can you put the command that you were trying to run into the issue if that's okay? Or is the error occurring when the application starts?

**Comment by SimonLab** [comment]
I think the error message is not linked to the command as it is displayed when the application is started. The app is working as usual but it uses all it's memory allocated (`Process running mem=767M(149.8%)`) for the free dyno (the standard and performance plan offer >=1Gb fo ram)
![image](https://user-images.githubusercontent.com/6057298/50825964-4ea15800-1332-11e9-8c13-8a66ab272cb0.png)

However I'm not sure yet if it's a normal use of the memory for the application.

**Comment by SimonLab** [comment]
@nelsonic I'm wondering if extracting the ETS table into another Heroku application will just move the memory error to this new application?

from https://devcenter.heroku.com/articles/heroku-postgres-plans it seems that heroku provides the `standard` plan for $50/month with no row limitation:
![image](https://user-images.githubusercontent.com/6057298/50828487-6d571d00-1339-11e9-8793-e5273fb8bd8c.png), would this be the cheapest solution if the memory problem persists?

**Comment by SimonLab** [comment]
I've tried to run `mix run priv/repo/add_lat_long_to_venue.exs` to update the database with the location for the venues but it makes the memory errors worth.

![image](https://user-images.githubusercontent.com/6057298/50829793-9c6f8d80-133d-11e9-932e-9d2cef170c47.png)

https://devcenter.heroku.com/articles/error-codes#r15-memory-quota-vastly-exceeded
![image](https://user-images.githubusercontent.com/6057298/50829866-c9bc3b80-133d-11e9-9089-9c3668edd162.png)
:confused:

**Comment by RobStallion** [comment]
I haven't had a chance to test this theory but I would assume that the error is happening because the file is too big to store on heroku, not because we are storing the file in ets. 

If this is the case then we could (possibly) solve this by streaming the postcode file rather than writing it to the project. 

Looking into possible solutions for this now

**Comment by nelsonic** [comment]
As @RobStallion says, we probably need to `delete` the Zip and Unzipped file on Heroku to ensure that the memory footprint of the App does not balloon.

**Comment by SimonLab** [comment]
From the application logs on Heroku, it looks like the file is already deleted: 
![image](https://user-images.githubusercontent.com/6057298/50831278-40f3ce80-1342-11e9-88f4-1aa88fc841fd.png)

Is this correct @RobStallion ?

**Comment by RobStallion** [comment]
@SimonLab It deletes the zip file but stores the ets file in the root file (so that we do not 'clone' the zip file from gh everytime)

**Comment by RobStallion** [comment]
I think that it might be a combo of both the csv file and the data being cached in ets. 

I ran the command `heroku run bash -a club-soda-guide-staging`
Once in, ran `ls -a`
Below is the list of files it logs...
![image](https://user-images.githubusercontent.com/15571853/50831754-ce83ee00-1343-11e9-8304-b343c99dcb7b.png)

I think heroku has deleted the csv file on it's own.
See [here](https://help.heroku.com/K1PPS2WM/why-are-my-file-uploads-missing-deleted) for more info (`"Each dyno boots with a clean copy of the filesystem from the most recent deploy."` <--- taken from the link)

Looking at the heroku logs @SimonLab posted above though it looks like the ets on its own may also have taken up too much memory. 

First it logs
```
2019-01-08T12:17:55.605167+00:00 app[web.1]: "Removed zip file. Storing postcodes with ets"
```

20 seconds later it logs the following...
```
2019-01-08T12:18:15.938482+00:00 heroku[web.1]: Process running mem=544M(105.7%)
2019-01-08T12:18:15.938632+00:00 heroku[web.1]: Error R14 (Memory quota exceeded)
2019-01-08T12:18:55.782821+00:00 heroku[web.1]: Process running mem=630M(123.2%)
```

This to me makes it seem like the caching process was running well for 20 seconds and then it exceeded the memory.

Looking into this more specifically...

heroku only allows for 512MB RAM...

<img src="https://user-images.githubusercontent.com/15571853/50832561-887c5980-1346-11e9-8671-294aa0ba1711.png" width="250px" />

but it appears that the postcode cache in ets is taking up a little over 6MB.

![image](https://user-images.githubusercontent.com/15571853/50832771-1d7f5280-1347-11e9-9068-6611febf751b.png)

Saw this by running the application in iex 

```
iex -S mix
```

and then running 

```
iex(1)> :observer.start
```

**Comment by RobStallion** [comment]
I think that we will have to either minimise the amount of memory the cache is taking up or remove the cache completely. 

Looking into possible solutions.

**Comment by RobStallion** [comment]
I think that (for now at least) it might be best to remove the ets cache from the application 😞.

Currently we are only using the cache for when we need to add lat long values to the database. I think that for this purpose it may be better to use the an api as we will not have to make many call and it not slow down the apps day to day use.

The ets solution may be better suited for a situation where lots of users are trying to search via a postcode as we would be able to quickly check the cache to see if the postcode was correct without having to make lots of individual api requests. 

That being said, it still could cause issues with heroku as the professional servers only offer 1GB RAM. This means that the postcode data as it currently stands will take up over half of the available ram.

**Comment by RobStallion** [comment]
@nelsonic Do you have any thoughts on the above? We could use the [postcode api](https://postcodes.io/) mentioned here https://github.com/club-soda/club-soda-guide/issues/129#issuecomment-448920300 to get the lat long values we need for the db.

This solution also means that we would not need to update the csv file with new postcodes when they are released.

**Comment by nelsonic** [comment]
@RobStallion using https://postcodes.io/ is viable provided it meets our service level (_response time and query volume_) requirements. That might be _easier_ than trying to prune the ETS table to reduce the cache size to only the postcodes we actually _need_ ... 🤔 

Pruning the ETS table to only store the postcodes with the same prefix as those where venues are located this would reduce the table down to around 10mb. 

Steps:
1. SELECT list of CS venues
2. Parse list of CS venue postcodes 
3. Wherever there is a CS venue, store the complete postcode list for that borough.
for example if the venue postcode is **MK40 1AY** https://www.clubsodaguide.com/search/venues
![image](https://user-images.githubusercontent.com/194400/50837729-4528e780-1354-11e9-995f-288dcb94af3d.png)
then store all the postcodes in **MK40** in the ETS lookup table.
This will mean that people who live in the Milton Keynes will find the "**Albero Lounge**" _fast_. 
4. When a person searches for their _complete_ postcode and we have an ETS cache _miss_, we may need to geocode it from the API, but we should store the _resulting_ geocode in ETS to speed up subsequent queries. 

We should _ideally_ have a `postcode_search` table in PostgreSQL which tracks which postcodes people are searching for. This will be _valuable_ data for the CS team to pitch membership to the pubs/venues in that postcode area if there are no results.

**Comment by jussi-club-soda** [comment]
Just having a quick look at this issue, not getting into the technical points, but am I reading @nelsonic correctly that we would only have postcode/coordinate data for locations where venues have already signed up from? What happens when a new one signs up somewhere else? We plan to cover all of UK, so potentially almost every postcode (first part of them) would be needed.

**Comment by nelsonic** [comment]
@jussi-club-soda _great_ question(s). I'm glad you have chimed in on the thread. 👍 
Rob's work has reduce the memory requirement for the _cache_ of the postcodes down to a manageable level so that _this_ issue "Memory Quota Exceeded" can be fixed.
We **100% agree** that the goal is to serve _all_ of the UK. 
The _beauty_ of the approach that Rob has taken is that each time a new venue is added the postcode will be in the database, so refreshing the ETS Postcode Cache (_which is what allows us to have **blazing fast** postcode lookups_) is simply a matter of "cycling" the app server as the cache generator scrip is part of the boot sequence. 
In the _meantime_ we will be working on _two_ additional enhancements:
**1.** _Proactive_ caching of the postcodes whenever a new venue is added. This will mean that if a venue is added with the **`BB9 7LJ`** postcode, _all_ of the surrounding postcodes in the **`BB9`** area will be cached for faster Geocoding and distance calculation. (_without having to pay Google $2k/year_)
**2.** Fallback to the postcodes.io API for postcodes that are _not_ in the ETS Cache (_for whatever reason_)

> Please note: the _other_ approach would be to just store the _whole_ database of postcodes _up-front_ and pay more for the Heroku dynos, but we _prefer_ to be **_memory-efficient_ up-front**, because we have gone down the route of "throw more memory at it" and it's basically an "unwinnable arms race" that just wastes time and a cash. 

Please let us know if this clarifies the approach taken and/or if you have any further questions.
Thanks! ✨

**Comment by RobStallion** [comment]
@SimonLab Can this be issue be closed now?

**Comment by SimonLab** [comment]
fixed with #323 
to continue with https://github.com/dwyl/phoenix-uk-postcode-finder-example
